### PR TITLE
[develop] Integrate Unified Workflow templater tool

### PR DESCRIPTION
**Pull request by christinaholtNOAA**
## DESCRIPTION OF CHANGES: 

This is the first time a Unified Workflow tool is being integrated into a UFS App. We are starting with the templater tool. It's a direct drop-in replacement for fill_jinja_template.py.

This contribution includes the following additional changes:
- Updates Externals.cfg to include workflow-tools as an external dependency repo (will be updated to point to a hash once the corresponding branch is merged)
- Updates module files for the workflow and the tasks to set PYTHONPATH to the necessary locations for uwtools.
- Allows the templater tool to do its own logging and error handling for clarity, instead of SRW callers.
- Lints the Python code perturbed by this integration in the spirit of stewardship.
- Adds a .pylintrc to help manage some of the project-wide settings that will be needed by SRW to lint appropriately.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 

- [x] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:

PR 237 in workflow-tools: https://github.com/ufs-community/workflow-tools/pull/237 (This has been merged and the hash updated in Externals.cfg)
For the linter to pass, we'll also need PR #788 to be merged in.

## DOCUMENTATION:

A small update was made.

## CHECKLIST

- [ ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

## LABELS (optional): 

A Code Manager needs to add the following labels to this PR: 
- [ ] Work In Progress
- [ ] bug
- [ ] enhancement
- [ ] documentation
- [ ] release
- [ ] high priority
- [ ] run_ci
- [ ] run_we2e_fundamental_tests
- [ ] run_we2e_comprehensive_tests
- [ ] Needs Cheyenne test 
- [ ] Needs Jet test 
- [ ] Needs Hera test 
- [ ] Needs Orion test 
- [ ] help wanted

## CONTRIBUTORS (optional): 

@fgabelmannjr @venitahagerty @WeirAE @elcarpenter @j-derrico

### CODE REVIEW

**Comment by christinaholtNOAA** [comment]
I am rebuilding and rerunning the fundamental intel tests on hera now after I merged in the latest develop. I'll report back the results.

**Comment by christinaholtNOAA** [comment]
Okay, Hera Intel fundamental tests still pass.

File: Externals.cfg
Comment (EdwardSnyder-NOAA) [inline Externals.cfg]: I do have to say it is a little confusing that the `uwtools` directory isn't under `sorc` directory, but if you check Externals.cfg, you'll see where it is downloaded. Given that the rest of the external packages are used within the model and `uwtools` is a python package for the workflow, I have no problem placing the directory under `ush/python_utils/uwtools`.

File: modulefiles/tasks/cheyenne/python_srw.lua
Comment (EdwardSnyder-NOAA) [inline modulefiles/tasks/cheyenne/python_srw.lua]: I'm curious if there is a reason why the scripts directory isn't at the end of this path? Doing that will cleanup the imports in the modified scripts under the `ush` directory to be `from templater import set_template` instead of `from scripts.templater import set_template`. It looks like PEP8 is okay with the latter but that's with subclasses and not a directory.

File: modulefiles/wflow_hera.lua
Comment (EdwardSnyder-NOAA) [inline modulefiles/wflow_hera.lua:10]: Are we adding this logic to the other `modulefiles/wflow_<platform>.lua` files? Also, are we going to change the printout in these script to `conda activate workflow_tools` instead of `conda activate regional_workflow`?

File: ush/generate_FV3LAM_wflow.py
Comment (EdwardSnyder-NOAA) [inline ush/generate_FV3LAM_wflow.py]: The next three changes in this script all cut the last character from the last word in the line.

File: modulefiles/tasks/cheyenne/python_srw.lua
Comment (christinaholtNOAA) [reply modulefiles/tasks/cheyenne/python_srw.lua]: That is certainly one way to go. It could still lead to the user not having any immediate clues where the `templater` is coming from though. My original thought was that leaving it here would allow us to be less invasive in the SRW integration when we want to import say a driver, which might end up at the same level as scripts under the workflow-tools repo.

I chatted briefly with the UW Team this morning about this general concern. I'm curious if we'd be good with just a comment (for now) in the uwtools-enabled Python scripts that points the user to where this comes from.  Here's what I'm thinking:

```
# These imports come from the uwtools package under ush/python_utils
from scripts.templater import set_template
```


We have a new team member coming in a matter of weeks that I'm confident could find a better solution than I have here. I believe that by the time the full integration of existing tools is done we'd have the more robust, informative solution in place. 

Would that be acceptable?

File: modulefiles/wflow_hera.lua
Comment (christinaholtNOAA) [reply modulefiles/wflow_hera.lua:10]: Ah! Yes. I missed the `workflow_tools` change for sure, and the other wflow module files. Great catches! Thanks.

File: Externals.cfg
Comment (christinaholtNOAA) [reply Externals.cfg]: The location was suggested by folks at EMC to make sure that our utility scripts stay under ush. Since we have a `python_utils` under there already, that seemed like the place. I could totally move it up one level, though.

Do you have any thoughts on aliasing it to `uwtools`? I could have it show up as a subdirectory named `workflow-tools` instead? It turned out a little wonky in my opinion that there is `ush/python_utils/uwtools/src/uwtools`. 

This was another choice on my part to decouple the name of our UW repo from the exact integration in SRW, but I could see how `uwtools` might be the wrong choice as `workflow-tools` is more readable as the subdirectory. Or even `workflow_tools`.

**Review by christinaholtNOAA — COMMENTED** [review]
@EdwardSnyder-NOAA Thanks for the review! I will get on those mods now!

File: Externals.cfg
Comment (EdwardSnyder-NOAA) [reply Externals.cfg]: The location under `python_utils` makes sense to me since it is a python tool. 

I think aliasing to `workflow-tools` or a spelled-out variant would be better as it is more readable and doesn't have the name listed twice in the path.

File: modulefiles/tasks/cheyenne/python_srw.lua
Comment (EdwardSnyder-NOAA) [reply modulefiles/tasks/cheyenne/python_srw.lua]: I'm fine with the suggested changes of leaving it as it is and adding a comment. I wasn't aware that other things like a driver could be imported, so I understand the reasoning. And I do think when a full integration of the tools occurs a cleaner solution will appear.

**Comment by christinaholtNOAA** [comment]
@natalie-perlin @MichaelLueken It's coming to my attention that these changes will need to be tested across all platforms, as we'd discussed in one of our recent dev meetings. Specifically, this PR changes the supported conda environment from regional_workflow to unified_workflow. 

We want to make sure it's available on the supported platforms for SRW.

I also don't know how to handle it on WCOSS, or how AQM will behave. Can @BenjaminBlake-NOAA and/or @chan-hoo take a look at testing this PR and providing the feedback necessary for running on WCOSS? 

Is NOAACLOUD still supported? Does it also need a workflow_tools environment?

File: .pylintrc
Comment (mkavulich) [inline .pylintrc:1]: Where does this file come from? Is there a particular reason to include so many custom rules rather than keeping the list small and specific to this project?

File: modulefiles/tasks/cheyenne/python_srw.lua
Comment (mkavulich) [inline modulefiles/tasks/cheyenne/python_srw.lua:5]: Where does this "workflow_tools" module come from? Is it from the unified workflow repository?

File: ush/generate_FV3LAM_wflow.py
Comment (mkavulich) [inline ush/generate_FV3LAM_wflow.py:120]: Does the new replacement script not have an equivalent quiet/verbose flag?

File: ush/generate_FV3LAM_wflow.py
Comment (mkavulich) [inline ush/generate_FV3LAM_wflow.py]: This f-string removal looks incorrect to me.
```suggestion
    logging.debug(f"Finished setting up debug file logging in {logfile}")
```

File: ush/generate_FV3LAM_wflow.py
Comment (mkavulich) [inline ush/generate_FV3LAM_wflow.py]: A merge seems to have gone wrong here, this is reverting changes from #778.

**Review by mkavulich — CHANGES_REQUESTED** [review]
A few questions and comments.

File: modulefiles/tasks/cheyenne/python_srw.lua
Comment (natalie-perlin) [reply modulefiles/tasks/cheyenne/python_srw.lua:5]: @mkavulich - `workflow_tools` is a conda environment built under the existing  **miniconda3/4.12.0** EPIC-managed installation, and contains conda/python packages as needed. It could be activated in the same way as `regional_workflow`.

File: .pylintrc
Comment (christinaholtNOAA) [reply .pylintrc:1]: This file is generated by pylint when running the `pylint --generate-rcfile`. Then I made a couple of changes to that default state for handling a couple of the issues specific to SRW.

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [reply ush/generate_FV3LAM_wflow.py:120]: It does have a quiet flag. I did not include it here because it logs information on its own that can be helpful in the log to see the flow of control through the program. 

It will, by default, log this info:

```
Running set_template with args: 
----------------------------------------------------------------------
----------------------------------------------------------------------
        outfile: None
 input_template: /path/to/input/template
    config_file:/path/to/config/file
   config_items: []
        dry_run: False
  values_needed: False
        verbose: False
          quiet: False
       log_file: /path/to/log/file
----------------------------------------------------------------------
----------------------------------------------------------------------
```

If it needs to be quiet, I can reimplement that.

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [reply ush/generate_FV3LAM_wflow.py]: Oops! Nice catch. I'll put it back.

**Review by christinaholtNOAA — COMMENTED** [review]
Thanks for the review @mkavulich

**Comment by chan-hoo** [comment]
> I also don't know how to handle it on WCOSS, or how AQM will behave. Can @BenjaminBlake-NOAA and/or @chan-hoo take a look at testing this PR and providing the feedback necessary for running on WCOSS?
> 
@christinaholtNOAA, I'll test this PR on WCOSS2 soon.

**Comment by BenjaminBlake-NOAA** [comment]
Thank you @chan-hoo !

File: modulefiles/tasks/cheyenne/python_srw.lua
Comment (natalie-perlin) [reply modulefiles/tasks/cheyenne/python_srw.lua:5]: @mkavulich -
    `workflow_tools `environment that was installed used the following yaml file from the workflow_tools repository:
    https://github.com/ufs-community/workflow-tools/blob/develop/environment.yaml
```
name: workflow_tools
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.9*
  - boto3=1.22*
  - black
  - f90nml=1.4*
  - jinja2=3.0*
  - jsonschema=4.17*
  - numpy=1.21*
  - pylint
  - pytest
  - pytest-cov=3.*
  - pyyaml=6.0*
  - tox
```
    
   NB: Requirement for python version is set to correspond to the same version in the default miniconda3/4.12.0 installation

**Comment by chan-hoo** [comment]
@christinaholtNOAA, I got the following error:
```
[chan-hoo.jeon@clogin07 ush] python3 generate_FV3LAM_wflow.py 
Traceback (most recent call last):
  File "generate_FV3LAM_wflow.py", line 37, in <module>
    from scripts.templater import set_template
ModuleNotFoundError: No module named 'scripts'
```
module list:
```
[chan-hoo.jeon@clogin07 modulefiles] module list

Currently Loaded Modules:
  1) craype-x86-rome     (H)   3) craype-network-ofi (H)   5) intel/19.1.3.304   7) set_pythonpath      9) wflow_wcoss2
  2) libfabric/1.11.0.0. (H)   4) envvar/1.0               6) python/3.8.6       8) core/rocoto/1.3.5
```

How can I resolve this?

**Comment by christinaholtNOAA** [comment]
@chan-hoo You will need to set the PYTHONPATH. I think you should be able to load the `modulefiles/set_pythonpath` module. You likely want to do this _after_ you have loaded your conda environment on WCOSS2 since that platform doesn't use conda to manage packages. 

The same will need to be done for run-time, but I'm not sure if it's okay to do that on WCOSS. Notice on lines 111-112 of `ush/load_modules_run_task.sh` I have that module loaded for all platforms except WCOSS2. Would you try it on WCOSS2 to see if those lines can be moved outside the if-statement?

There's a chance we'd need to do that _after_ loading the python environment module though to make sure the right PYTHONPATH is priority order is set.

**Comment by chan-hoo** [comment]
@christinaholtNOAA
```
[chan-hoo.jeon@clogin07 ufs-srweather-app] source versions/run.ver
[chan-hoo.jeon@clogin07 ufs-srweather-app] module use modulefiles/
[chan-hoo.jeon@clogin07 ufs-srweather-app] module load wflow_wcoss2
[chan-hoo.jeon@clogin07 ufs-srweather-app] module list

Currently Loaded Modules:
  1) craype-x86-rome     (H)   3) craype-network-ofi (H)   5) intel/19.1.3.304   7) set_pythonpath      9) wflow_wcoss2
  2) libfabric/1.11.0.0. (H)   4) envvar/1.0               6) python/3.8.6       8) core/rocoto/1.3.5

[chan-hoo.jeon@clogin07 ufs-srweather-app] echo $PYTHONPATH
modulefiles/../ush/python_utils/workflow-tools/src:modulefiles/../ush/python_utils/workflow-tools:/apps/prod/python-modules/3.8.6/intel/19.1.3.304/lib/python3.8/site-packages
```
PYTHONPATH seems to be set as intended. Should we put the full paths? Is there any way to put the full path rather than `splitFileName`?
```
local mod_path, mod_file = splitFileName(myFileName())
local uwtools_scripts_path =  pathJoin(mod_path, "/../ush/python_utils/workflow-tools")
```

**Comment by chan-hoo** [comment]
@christinaholtNOAA, I tried adding the full (absolute) path to the modulefiles directory manually and the workflow was generated correctly. However, I have no idea how to get a full path. Please let me know how to do it.

**Comment by christinaholtNOAA** [comment]
@chan-hoo I'm curious if it's related to how you load the module. If you go with 

```
module use /full/path/to/ufs-srweather-app/modulefiles
```

Will that do it?

**Comment by chan-hoo** [comment]
@christinaholtNOAA, it worked if I put a full path as you said. However, in this case, we should put a full path whenever we generates a workflow on a new terminal window. Please update `set_pythonpath.lua` to guarantee a full path even though we type 'module use modulefiles' if you have a solution.

File: ush/load_modules_run_task.sh
Comment (chan-hoo) [inline ush/load_modules_run_task.sh:106]: @christinaholtNOAA, as you mentioned, please update this part like:
```
module use "${default_modules_dir}"
if [ "${machine}" != "wcoss2" ]; then
  module load "${BUILD_MOD_FN}" || print_err_msg_exit "\
  Loading of platform- and compiler-specific module file (BUILD_MOD_FN)
for the workflow task specified by task_name failed:
  task_name = \"${task_name}\"
  BUILD_MOD_FN = \"${BUILD_MOD_FN}\""
fi
module load set_pythonpath || print_err_msg_exit "\
  Loading the module to set PYTHONPATH for workflow-tools failed."
```

File: ush/create_aqm_rc_file.py
Comment (chan-hoo) [inline ush/create_aqm_rc_file.py]: @christinaholtNOAA, I got the following error in 'run_fcst' of AQM:
```
Traceback (most recent call last):
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/create_aqm_rc_file.py", line 179, in <module>
    create_aqm_rc_file(
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/create_aqm_rc_file.py", line 13
5, in create_aqm_rc_file
    set_template(
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/scripts/templater.py", line 105, in set_template
    user_args = parse_args(argv)
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/scripts/templater.py", line 79, in parse_args
    return parser.parse_args(argv)
  File "/apps/spack/python/3.8.6/intel/19.1.3.304/pjn2nzkjvqgmjw4hmyz43v5x4jbxjzpk/lib/python3.8/argparse.py", 
line 1768, in parse_args    args, argv = self.parse_known_args(args, namespace)
  File "/apps/spack/python/3.8.6/intel/19.1.3.304/pjn2nzkjvqgmjw4hmyz43v5x4jbxjzpk/lib/python3.8/argparse.py", 
line 1800, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)  File "/apps/spack/python/3.8.6/intel/19.1.3.304/pjn2nzkjvqgmjw4hmyz43v5x4jbxjzpk/lib/python3.8/argparse.py", 
line 1841, in _parse_known_args    option_tuple = self._parse_optional(arg_string)
  File "/apps/spack/python/3.8.6/intel/19.1.3.304/pjn2nzkjvqgmjw4hmyz43v5x4jbxjzpk/lib/python3.8/argparse.py", 
line 2128, in _parse_optional
     if not arg_string[0] in self.prefix_chars:
TypeError: '_TemporaryFileWrapper' object is not subscriptable
+ exregional_run_fcst.sh[587]: print_err_msg_exit 'Call to function to create an aqm.rc file for the current
cycle'\''s (cdate) run directory (DATA) failed:
```

**Comment by christinaholtNOAA** [comment]
@chan-hoo In response to the request for the full module path, it seems that this is an issue only on WCOSS2 and since I don't have access, I can't debug the different behavior of the LUA module. For example, when I run as you did on Hera, the full path is used:

```
$ echo $PYTHONPATH

$ module use modulefiles
$ module load set_pythonpath
$ echo $PYTHONPATH
/scratch2/BMC/wrfruc/cholt/srw_work/uw_tools/ufs-srweather-app/modulefiles/../ush/python_utils/workflow-tools/src:/scratch2/BMC/wrfruc/cholt/srw_work/uw_tools/ufs-srweather-app/modulefiles/../ush/python_utils/workflow-tools
```

File: .pylintrc
Comment (christinaholtNOAA) [reply .pylintrc:1]: I supposed I never responded to the second half of this comment. Sorry. If we keep all of them, it protects us against new versions of pylint changing defaults on us.

File: ush/generate_FV3LAM_wflow.py
Comment (mkavulich) [reply ush/generate_FV3LAM_wflow.py:120]: I was actually asking about the opposite. The previous script had increased verbosity if the "quiet" flag was not provided, and this was toggled in generate_FV3LAM_wflow.py with the `debug` flag. Is there an equivalent flag to `debug` that we can pass to this new script to get increased verbosity?

File: ush/create_aqm_rc_file.py
Comment (christinaholtNOAA) [reply ush/create_aqm_rc_file.py]: I believe I missed calling `set_template` with `tmpfile.name` instead of `tmpfile`. I've pushed the change that will hopefully take care of that.

File: ush/generate_FV3LAM_wflow.py
Comment (mkavulich) [reply ush/generate_FV3LAM_wflow.py:120]: It looks like this does not exist. Resolving this conversation.

**Review by mkavulich — APPROVED** [review]
All my comments have been addressed

File: ush/generate_FV3LAM_wflow.py
Comment (christinaholtNOAA) [reply ush/generate_FV3LAM_wflow.py:120]: I will add an Issue to UW to add some additional debug output. Our `set_config` does have this capability, and it's a good one.

File: ush/load_modules_run_task.sh
Comment (chan-hoo) [inline ush/load_modules_run_task.sh]: @christinaholtNOAA, this line should move out of the if-statement (to Ln 104).

File: ush/create_aqm_rc_file.py
Comment (chan-hoo) [reply ush/create_aqm_rc_file.py]: @christinaholtNOAA, I got another error in AQM:
```
Traceback (most recent call last):
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/create_aqm_rc_file.py", line 179
, in <module>
    create_aqm_rc_file(
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/create_aqm_rc_file.py", line 135
, in create_aqm_rc_file
    set_template(
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/scripts/templater.py", line 119, in set_template
    cfg = setup_config_obj(user_args, log_name=log.name)
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/scri
pts/templater.py", line 88, in setup_config_obj    config_type = cli_helpers.get_file_type(user_args.config_file)
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/src/uwtools/utils/cli_helpers.py", line 33, in get_file_type
    raise ValueError(msg)
ValueError: Bad file suffix -- . Cannot determine file type!
```
You can check this error on Hera too. The sample script of the community mode of AQM is available on Hera.

**Comment by christinaholtNOAA** [comment]
It appears that the functional test is failing because data was removed from AWS. This date only goes out to 17 z files now. Thoughts on how to "fix" this failure?

Switch to a different date that does have data?

**Comment by christinaholtNOAA** [comment]
@chan-hoo @EdwardSnyder-NOAA Have I addressed your comments sufficiently, or are there still outstanding issues I can work to resolve?

**Comment by MichaelLueken** [comment]
@christinaholtNOAA Similar to what was done for the task-based `python_srw.lua` modulefiles, should the `wflow_*.lua` modulefiles also be updated to replace `regional_workflow` with `workflow_tools`?

**Comment by chan-hoo** [comment]
@christinaholtNOAA, I think you missed my comment above:
```
```
Traceback (most recent call last):
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/create_aqm_rc_file.py", line 179
, in <module>
    create_aqm_rc_file(
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/create_aqm_rc_file.py", line 135
, in create_aqm_rc_file
    set_template(
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/scripts/templater.py", line 119, in set_template
    cfg = setup_config_obj(user_args, log_name=log.name)
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/scri
pts/templater.py", line 88, in setup_config_obj    config_type = cli_helpers.get_file_type(user_args.config_file)
  File "/lfs/h2/emc/lam/noscrub/chan-hoo.jeon/srw_pr_test/ufs-srweather-app/ush/python_utils/workflow-tools/src/uwtools/utils/cli_helpers.py", line 33, in get_file_type
    raise ValueError(msg)
ValueError: Bad file suffix -- . Cannot determine file type!
```
```

File: ush/create_aqm_rc_file.py
Comment (christinaholtNOAA) [reply ush/create_aqm_rc_file.py]: I'm sorry! I did miss this one! I will tackle that now.

**Comment by EdwardSnyder-NOAA** [comment]
> @chan-hoo @EdwardSnyder-NOAA Have I addressed your comments sufficiently, or are there still outstanding issues I can work to resolve?

Yes, you have. Approving on my end.

**Review by EdwardSnyder-NOAA — APPROVED** [review]

**Comment by christinaholtNOAA** [comment]
@chan-hoo I pushed a change this morning to hopefully address the failure you were seeing.

**Review by MichaelLueken — APPROVED** [review]
@christinaholtNOAA These changes look good to me!  I was able to clone your branch on Tier-1 platforms (except for Gaea, as the SRW App won't compile or run following the system upgrade) and run the WE2E fundamental tests.

I did note that loading the `regional_workflow` conda environment and attempting to run the fundamental tests would cause some of the tests to fail.  Loading the `workflow_tools` conda environment, all tests seem to pass successfully.  I would recommend that the modulefiles/wflow_*.lua files be updated to replace `regional_workflow` with `workflow_tools` (similar to the change made in `ush/load_modules_wflow.sh`).

Please note that the `grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16` fundamental test is failing on Cheyenne GNU.  Running the test on the current HEAD of develop shows that the test is failing there as well, so I won't hold this PR up for this reason (issue #817 has been opened to document this failure).

**Comment by christinaholtNOAA** [comment]
@MichaelLueken I will try to get to that change shortly. Thanks for the suggestion.

**Review by MichaelLueken — COMMENTED** [review]
@christinaholtNOAA In addition to changing the `regional_workflow` to `workflow_tools` in the `wflow_*.lua` modulefiles, please change `modulefiles/tasks/jet/run_vx.local.lua` to use:

`load("python_srw")`

The `grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16` fundamental test is failing on Jet because the necessary conda environment isn't available (the test is failing because it is unable to find f90nml).

**Review by chan-hoo — APPROVED** [review]
@christinaholtNOAA, a test run for AQM was completed successfully on wcoss2.

**Comment by MichaelLueken** [comment]
@christinaholtNOAA Given @chan-hoo's approval, I will move forward running the Jenkins tests on this work.  The modifications to the `wflow_*.lua` modulefiles and `modulefiles/tasks/jet/run_vx.local.lua` file shouldn't affect the automated tests.  This will also make sure that the automated tests are properly set up to run with the UW templater too.

**Comment by MichaelLueken** [comment]
@christinaholtNOAA The automated Jenkins tests encountered failures:

Cheyenne GNU - the `nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km` test failed (detailed in issue https://github.com/ufs-community/ufs-srweather-app/issues/806).

Gaea - SRW won't build following the latest maintenance on the machine. Will be turning off Gaea testing in https://github.com/ufs-community/ufs-srweather-app/pull/799 until the UFS-WM has merged the new Gaea modulefile changes.

Hera Intel - the `grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test failed (detailed in issue https://github.com/ufs-community/ufs-srweather-app/issues/805).

I saw that the there were two failures in the Hera GNU test - `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` and `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`. These tests shouldn't be failing, so I have resubmitted them. I will let you know if they continue to fail.

On Jet, the `grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test failed in `run_post_mem000_f003` with the message:

`forrtl: severe (41): insufficient virtual memory`

I will resubmit the Jet tests and hopefully this test will pass.

I will also run the Jenkins tests manually on Orion, since the machine requires git/2.28.0 to be loaded, but since git/2.28.0 isn't available on Hercules and Hercules and Orion are the same machine, the git/2.28.0 module load is constantly removed from the role-epic-ps account's .bashrc file. I will let you know if there are any failures.

**Comment by MichaelLueken** [comment]
@christinaholtNOAA The manual testing of the Jenkins coverage tests on Orion successfully passed.  However, the same test that failed on Jet, `grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta`, has failed again.  This time, it has failed in `run_post_mem000_f001` with the following message:

```
srun: error: k4: task 13: Killed
srun: launch/slurm: _step_signal: Terminating StepId=28933725.0
slurmstepd: error: *** STEP 28933725.0 ON k4 CANCELLED AT 2023-06-05T15:36:57 ***
```

The test can be found - `/mnt/lfs4/HFIP/hfv3gfs/role.epic/jenkins/workspace/fs-srweather-app_pipeline_PR-793/expt_dirs/grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta`.

**Comment by christinaholtNOAA** [comment]
@MichaelLueken I saw that the test that was failing on Jet yesterday now seems to be passing. I also saw that running that manual test on Jet passed for me yesterday. 

Is there any update on the status of the other failures? It seems we may want to look more into random failures for the post if we keep seeing those creep in.

**Comment by MichaelLueken** [comment]
@christinaholtNOAA The failures are on Gaea (the SRW App won't compile following the maintenance on the machine that changed the available libraries on the machine), Orion (the necessary git/2.28.0 module isn't being loaded in the role-epic-ps account, causing the Jenkins pipeline to fail to clone the CCPP), and an issue on Hera GNU (the `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` test is failing).  I have manually run the Orion tests and all tests passed.  I will rerun the Hera GNU test that failed, then move forward with merging once it has passed.

**Comment by MichaelLueken** [comment]
@christinaholtNOAA The Hera GNU tests successfully completed during the rerun, so I will now move forward with merging this PR.

Additionally, it is important to note that the `grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test successfully passed with this PR.  I won't close issue #805 right away, but will perform additional testing before closing the issue as resolved.  The `nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km` is still failing in this PR.  So, that issue is still open.  I haven't been looking into these failures either, I've been focusing on fixing the issues with the updated WM hashes.

**Pull request merged**
